crates/starknet-devnet/tests/test_blocks_on_demand.rs

```diff
@@ -0,0 +1,175 @@
+
+        devnet.create_block().await.unwrap();
+
+        let pending_block = devnet.get_pending_block_with_tx_hashes().await.unwrap();
+        let latest_block = devnet.get_latest_block_with_tx_hashes().await.unwrap();
+
+        // querying pending block in normal mode should default to the latest block
+        assert_eq!(pending_block, latest_block);
+    }
+
+    #[tokio::test]
+    async fn blocks_on_demand_invoke_and_call() {
+        let devnet =
+            BackgroundDevnet::spawn_with_additional_args(&["--blocks-on-demand"]).await.unwrap();
+
+        let (signer, account_address) = devnet.get_first_predeployed_account().await;
+        let predeployed_account = Arc::new(SingleOwnerAccount::new(
+            devnet.clone_provider(),
+            signer.clone(),
+            account_address,
+            constants::CHAIN_ID,
+            ExecutionEncoding::New,
+        ));
+
+        let (contract_class, casm_class_hash) =
+            get_simple_contract_in_sierra_and_compiled_class_hash();
+
+        // declare the contract
+        let declaration_result = predeployed_account
+            .declare(Arc::new(contract_class), casm_class_hash)
+            .max_fee(FieldElement::from(1e18 as u128))
+            .send()
+            .await
+            .unwrap();
+
+        devnet.create_block().await.unwrap();
+
+        // deploy the contract
+        let contract_factory =
+            ContractFactory::new(declaration_result.class_hash, predeployed_account.clone());
+        let initial_value = FieldElement::from(10_u32);
+        let ctor_args = vec![initial_value];
+        contract_factory
+            .deploy(ctor_args.clone(), FieldElement::ZERO, false)
+            .max_fee(FieldElement::from(1e18 as u128))
+            .send()
+            .await
+            .unwrap();
+
+        devnet.create_block().await.unwrap();
+
+        // generate the address of the newly deployed contract
+        let contract_address = get_udc_deployed_address(
+            FieldElement::ZERO,
+            declaration_result.class_hash,
+            &starknet_rs_core::utils::UdcUniqueness::NotUnique,
+            &ctor_args,
+        );
+
+        let increment = FieldElement::from(5_u32);
+        let contract_invoke = vec![Call {
+            to: contract_address,
+            selector: get_selector_from_name("increase_balance").unwrap(),
+            calldata: vec![increment, FieldElement::ZERO],
+        }];
+
+        let invoke_result = predeployed_account
+            .execute(contract_invoke.clone())
+            .max_fee(FieldElement::from(1e18 as u128))
+            .send()
+            .await
+            .unwrap();
+
+        devnet.create_block().await.unwrap();
+
+        assert_tx_successful(&invoke_result.transaction_hash, &devnet.json_rpc_client).await;
+        assert_eq!(
+            get_contract_balance(&devnet, contract_address).await,
+            initial_value + increment
+        );
```

Comment:
Can we also assert here that all of this resulted in one block? and that the block contains only the tx hashes corresponding to declare, deploy, invoke

Reply:
I'm not sure if you can declare, deploy, and invoke with in one block in Starknet

Reply:
I think this is a very important part of the pending state/block functionality, this needs to be discussed in more detail

Reply:
discussed on a call I added additional invoke here 7f5f5c7db5a702c5e0e3e776b1eddbacf6d6c500